### PR TITLE
Parse: update readings enc_key + rolling_code; Anpassung autocreate

### PR DESCRIPTION
**Pull request by habeIchVergessen**

### CODE REVIEW

**Comment by viegener** [comment]
This is quite some extensive changes, and without explanation hard to assess. Can this be broken down into smaller pieces?

File: Somfy/10_SOMFY.pm [hunk 1]
```diff
@@ -524,6 +523,7 @@ sub SOMFY_Parse($$) {
 	if ($ioType eq "SIGNALduino") {
 		my $encData = substr($msg, 2);
 		return "Somfy RTS message format error!" if ($encData !~ m/A[0-9A-F]{13}/);
+		return "Somfy RTS message format error (length)!" if (length($encData) != 14);
```
Comment (habeIchVergessen) [inline Somfy/10_SOMFY.pm:526]: additional validation of received data

File: Somfy/10_SOMFY.pm [hunk 2]
```diff
@@ -391,18 +391,28 @@ sub SOMFY_SendCommand($@)
...
 
-	my $enckey = uc(ReadingsVal($name, "enc_key", "A0"));
-	my $rollingcode = uc(ReadingsVal($name, "rolling_code", "0000"));
-
 	if($command eq "XX") {
 		# use user-supplied custom command
 		$command = $args[1];
 	}
 
+	# increment before sending
```
Comment (habeIchVergessen) [inline Somfy/10_SOMFY.pm:399]: SOMFY_SendCommand: just changed the order of operation. don't increase enc key and rolling code in the end of last sending.

File: Somfy/10_SOMFY.pm [hunk 3]
```diff
@@ -571,23 +573,22 @@ sub SOMFY_Parse($$) {
...
 
-	} else {
-		# rolling code and enckey
-		my $rolling = substr($msg, 6, 4);
-		my $encKey = substr($msg, 2, 2);
-		
-		Log3 $hash, 1, "SOMFY Unknown device $address ($encKey $rolling), please define it";
-		return "UNDEFINED SOMFY_$address SOMFY $address $encKey $rolling";
+		# return list of affected devices
+		# otherwise goto autocreate
```
Comment (habeIchVergessen) [inline Somfy/10_SOMFY.pm:586]: slightly different handling for device detecting

**Comment by habeIchVergessen** [comment]
added comments in view of changes

**Review by viegener — CHANGES_REQUESTED** [review]
I do not understand the reasoning for the change of order in operation, it would mean that now the reading rolling code will contain the latest (already used) rolling code and not the next one that will be used. Why is this needed to support Singalduino? It will affect existing users.

**Comment by habeIchVergessen** [comment]
at least there is no difference between increasing rolling code + encryption key before or after sending.
for testing purpose i used a minuCUL for sending and received the message with signal duino in the same fhem instance. this constalation results in unchanged values, because the processing of received message decreasing rolling code and encryption key immediately.

**Comment by viegener** [comment]
The problem is a different one - It is the issue that for existing users the behavior will be changed and the rolling code will not match anymore, since now the sematics of the current reading is changed.
The change seems required since the message received and the message sent are the same. I think there are three options: 
- Rolling_codes are not updated from parsed/received messages
- Rolling codes are increased by one when receiving a message in aprse function
- sent messages are put (via AddDuplicate) to the duplicate list and therefore ignored when receiving the same message 

The second solution would be least invasive and not changing existing behavior

**Pull request closed without merging**